### PR TITLE
🌱 Avoid MachineHealthCheck to return early on patch errors

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:
With this change, MachineHealtchCheck tries to patch all unhealthy Machines and return at the end if there is error.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3712

### CODE REVIEW

File: controllers/machinehealthcheck_controller.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:267]: We probably want to `continue` here, otherwise the event below is registered

**Review by vincepri — APPROVED** [review]
/approve
/assign @CecileRobertMichon 
/milestone v0.3.10

**Comment by vincepri** [comment]
@sedefsavas Can we change the commit message to match the PR's title?

**Comment by ncdc** [comment]
LGTM. Is there any way we can somehow write a test for this without it being extremely messy?

**Comment by sedefsavas** [comment]
@ncdc I am looking at a way to make patch fail on one of the MHC target machines to test this.

**Comment by fabriziopandini** [comment]
lgtm
 
> I am looking at a way to make patch fail on one of the MHC target machines to test this.

What if we break down markUnhealty (and markHealthy) in separated func, so it will be easier to create scenarios where e.g. an unhealthy machine does not exists in the cluster.

**Comment by vincepri** [comment]
Changed the base branch to be release-0.3, we'll have to forward-port these changes to v0.4 later

**Comment by vincepri** [comment]
@ncdc @fabriziopandini @CecileRobertMichon This is probably ready for another pass

**Comment by sedefsavas** [comment]
This should be ready to go, PTAL.

cc @vincepri @ncdc

**Comment by vincepri** [comment]
LGTM, although this seems it's going to conflict with #3606 - is it ok if we wait for that one to go in first and then rebase?

**Comment by sedefsavas** [comment]
Ok, putting a hold to this then.

/hold
until #3606 is merged

**Comment by sedefsavas** [comment]
`pull-cluster-api-apidiff-release-0-3` is failing due to an optional field added to a recently added `NewTestClusterCacheTracker ` that is used in tests.

We can ignore this.

**Pull request merged**
